### PR TITLE
Use traitlets>=4.1 APIs

### PR DESCRIPTION
**Pull request by SylvainCorlay**
`_foo_changed` -> `@observe('foo')` + signature change.
`_bar_default` -> `@default('bar')`

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
@takluyver there will be less deprecation warnings.

File: ipykernel/kernelbase.py [hunk 1]
```diff
@@ -345,19 +343,19 @@ def send_response(self, stream, msg_or_type, content=None, ident=None,
         """
         return self.session.send(stream, msg_or_type, content, self._parent_header,
                                  ident, buffers, track, header, metadata)
-    
+
     def init_metadata(self, parent):
         """Initialize metadata.
-        
+
```
Comment (takluyver) [inline ipykernel/kernelbase.py]: You know what we say about whitespace changes ;-)

File: ipykernel/comm/comm.py [hunk 2]
```diff
@@ -21,15 +21,22 @@ class Comm(LoggingConfigurable):
     shell = Instance('IPython.core.interactiveshell.InteractiveShellABC',
                      allow_none=True)
     kernel = Instance('ipykernel.kernelbase.Kernel')
-    def _kernel_default(self):
+
+    @default('kernel')
+    def _default_kernel(self):
```
Comment (takluyver) [inline ipykernel/comm/comm.py]: This feels redundant - since anonymous functions aren't really a thing in Python, we're likely going to keep naming them like this anyway. So the decorator is just duplicating information. If that's how traitlets works now, though, I guess that's what we'll do. :neutral_face:

Comment (SylvainCorlay) [inline ipykernel/comm/comm.py]: The new API enabled a lot of simplifications in ipywidgets but for simple things like this, it is true that the net contribution is positive in terms of lines added/removed. ..

File: ipykernel/kernelbase.py (hunk 1 above)
Comment (SylvainCorlay) [inline ipykernel/kernelbase.py]: Yeah but there are *good* whitespace changes and *bad* whitespace changes... ;)

File: ipykernel/comm/comm.py (hunk 2 above)
Comment (minrk) [inline ipykernel/comm/comm.py]: Yeah, I found this when doing the update for JupyterHub. For more advanced uses, the changes are really nice, but for most of our simple use it really is not an improvement. Using traitlets for simple config has become more tedious than it used to be.

File: ipykernel/kernelapp.py [hunk 3]
```diff
@@ -16,17 +16,16 @@
...
-    InteractiveShellApp, shell_flags, shell_aliases
+        InteractiveShellApp, shell_flags, shell_aliases
 )
 from IPython.utils import io
 from ipython_genutils.path import filefind, ensure_dir_exists
-from traitlets import (
-    Any, Instance, Dict, Unicode, Integer, Bool, DottedObjectName, Type,
-)
+from traitlets import (Any, Instance, Dict, Unicode, Integer, Bool,
+                       DottedObjectName, Type, default)
```
Comment (minrk) [inline ipykernel/kernelapp.py]: I won't ask you to revert these changes, but FWIW, I prefer the form of all of these before this PR to after.

File: setup.py [hunk 4]
```diff
@@ -81,7 +81,7 @@
 setuptools_args = {}
 install_requires = setuptools_args['install_requires'] = [
     'ipython>=4.0.0',
-    'traitlets>=4.1.0',
+    'traitlets>=4.2.0',
```
Comment (minrk) [inline setup.py]: What in this PR changes the minimum version to 4.2? Aren't all these APIs in 4.1?

**Comment by minrk** [comment]
A few comments inline, but I only care about the traitlets 4.2 question before merging.

File: setup.py (hunk 4 above)
Comment (SylvainCorlay) [inline setup.py]: That is right, but traitlets 4.2 has (unrelated) bug fixes and is out there. I can revert that if you want.

File: ipykernel/kernelapp.py (hunk 3 above)
Comment (SylvainCorlay) [inline ipykernel/kernelapp.py]: I can revert this when I fix the other part.

**Comment by SylvainCorlay** [comment]
@minrk I updated/squashed as per your comments.

**Comment by minrk** [comment]
Thanks!
